### PR TITLE
RavenDB-15069 Better error when using features from a grater major version

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-15069

### CODE REVIEW

File: src/Raven.Server/RavenServerStartup.cs [hunk 1]
```diff
@@ -246,6 +239,35 @@ private async Task RequestHandler(HttpContext context)
...
+                    $"RavenDB does not support interaction between Client API major version 3 and Server version {RavenVersionAttribute.Instance.MajorVersion} when major version does not match. " +
+                    $"Client: {version}. Server: {RavenVersionAttribute.Instance.AssemblyVersion}",
+                    e);
+                return;
+            }
+
+            var nVersion = char.GetNumericValue(version, 0);
+            if (nVersion > RavenVersionAttribute.Instance.MajorVersion)
+            {
+                e = new ClientVersionMismatchException(
```
Comment (ppekrol) [inline src/Raven.Server/RavenServerStartup.cs]: What will happen if we throw e.g. 'IndexDoesNotExistException' when Client version is 5.0.0.50?

Comment (haludi) [reply src/Raven.Server/RavenServerStartup.cs]: So we will get a warning about the version mismatch.

Comment (ppekrol) [reply src/Raven.Server/RavenServerStartup.cs]: Which is incorrect? Right?

File: src/Raven.Server/RavenServerStartup.cs [hunk 2]
```diff
@@ -246,6 +241,35 @@ private async Task RequestHandler(HttpContext context)
...
+                    e);
+            }
+            else if (HasInvalidCommandTypeException(e))
+            {
+                RequestRouter.CheckClientVersionAndWrapException(version, ref e);
+            }
+
+            static bool HasInvalidCommandTypeException(Exception e)
+            {
+                if (e.GetType() == typeof(BatchRequestParser.InvalidCommandTypeException))
```
Comment (ayende) [inline src/Raven.Server/RavenServerStartup.cs]:
```suggestion
                if (e is BatchRequestParser.InvalidCommandTypeException)
```

File: src/Raven.Server/Routing/RequestRouter.cs [hunk 3]
```diff
@@ -223,18 +223,28 @@ public static bool TryGetClientVersion(HttpContext context, out Version version)
...
+            if(CheckClientVersionAndWrapException(clientVersion, ref innerException) == false)
+                throw innerException;
+        }
+
+        public static bool CheckClientVersionAndWrapException(Version clientVersion, ref Exception innerException)
+        {
+            var currentServerVersion = RavenVersionAttribute.Instance;
+
+            if (currentServerVersion.MajorVersion == clientVersion.Major &&
+                currentServerVersion.BuildVersion >= clientVersion.Revision &&
```
Comment (ayende) [inline src/Raven.Server/Routing/RequestRouter.cs:238]: Why are you comparing build version to revision here?

File: src/Raven.Server/Routing/RequestRouter.cs [hunk 4]
```diff
@@ -223,18 +223,28 @@ public static bool TryGetClientVersion(HttpContext context, out Version version)
...
+                currentServerVersion.BuildVersion != ServerVersion.DevBuildNumber &&
+                (clientVersion.Revision < 40 || clientVersion.Revision >= 50))
```
Comment (ayende) [inline src/Raven.Server/Routing/RequestRouter.cs]: Why this line?

File: src/Raven.Server/Routing/RequestRouter.cs (hunk 3 above)
Comment (haludi) [reply src/Raven.Server/Routing/RequestRouter.cs:238]: There are two different classes.
The `currentServerVersion` is `RavenVersionAttribute`
And the `clientVersion` is System.Version.
So the `RavenVersionAttribute.BuildVersion` is equivalent to `System.Version.Revision`

File: src/Raven.Server/Routing/RequestRouter.cs (hunk 4 above)
Comment (haludi) [reply src/Raven.Server/Routing/RequestRouter.cs]: If I understand it right it is to check if the client unofficially built (in a development environment)

File: src/Raven.Server/Documents/Handlers/BatchRequestParser.cs [hunk 5]
```diff
@@ -1024,5 +1014,9 @@ private static void ThrowUnexpectedEndOfStream()
             throw new EndOfStreamException();
         }
 
+        public class InvalidCommandTypeException : Exception
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/BatchRequestParser.cs]: We have dedicated namespace for exceptions. Raven.Server.Exceptions. Move it there.

File: src/Raven.Server/RavenServerStartup.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Server/RavenServerStartup.cs]: +1

File: src/Raven.Server/RavenServerStartup.cs [hunk 6]
```diff
@@ -246,6 +241,35 @@ private async Task RequestHandler(HttpContext context)
...
+                    return true;
+
+                if (e is AggregateException ae)
```
Comment (ppekrol) [inline src/Raven.Server/RavenServerStartup.cs:267]: Is this a real case? Do we have aggregate exceptions anywhere? The .Any is not a cheap operaration.

File: src/Raven.Server/Routing/RequestRouter.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Routing/RequestRouter.cs]: You are excluding only 40 .. 50 here?

Comment (haludi) [reply src/Raven.Server/Routing/RequestRouter.cs]: This is the range of v4.x other versions will be check by the first condition

Comment (ayende) [reply src/Raven.Server/Routing/RequestRouter.cs]: Problem is that the 6.0 client will be accepted here, accidently.

Comment (haludi) [reply src/Raven.Server/Routing/RequestRouter.cs]: How? The major is not equal?

Comment (ayende) [reply src/Raven.Server/Routing/RequestRouter.cs]: I don't like that line. It make no sense to me. 
The dev version we use is either 40 or 50 for dev, no? In that case, check for those using equality, please.

File: src/Raven.Server/RavenServerStartup.cs (hunk 6 above)
Comment (haludi) [reply src/Raven.Server/RavenServerStartup.cs:267]: Yes in the bulk insert it is. I can convert it to `foreach`

File: src/Raven.Server/Documents/Handlers/BatchRequestParser.cs [hunk 7]
```diff
@@ -950,46 +951,36 @@ private static unsafe CommandType GetCommandType(JsonParserState state, JsonOper
...
-                    
-                    ThrowInvalidProperty(state, ctx);
-                    return CommandType.None;
-                
-                default:
-                    ThrowInvalidProperty(state, ctx);
-                    return CommandType.None;
+                    break;
             }
+            throw new InvalidCommandTypeException("Invalid command type: " + ctx.AllocateStringValue(null, state.StringBuffer, state.StringSize));
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/BatchRequestParser.cs]: Better to move it to a dedicated method - like we did with `ThrowInvalidProperty`. The reason is slightly better performance - parsing is our hot path.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**